### PR TITLE
listener: introduce tcp listener and udp listener interface

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:

Split from #15127

This PR introduces the interface at the least changes.
Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Additional Description:
Follow up
1. Move the active tcp listener and udp out of conn_handler_impl class and put in separate files.
2. Use mock to unit test tcp listener and udp listener to increase deterministic and avoid construct listener config and addListener() in each test case. Part of #15239
Risk Level: LOW
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lambdai** [comment]
This is the simplified code.

The interfaces:
ConnectionHandler: used by worker to add/delete listener from config. The listener can be udp listener or tcp listener.
TcpListener: The connection/tcp socket view of the listener. A tcp connection need rebalancer  but doesn't need an addListener() (I admit TcpListener interface currently allow to access addListener(). That could be addressed later)
Udplistener: similar to TcpListener but for udp listener

**Comment by lambdai** [comment]
The clang tidy error should be false alarm. Add ENVOY_BUG in that condition

File: source/server/connection_handler_impl.h [hunk 1]
```diff
@@ -332,12 +346,20 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
...
     bool connected_{false};
   };
-
   using ActiveTcpListenerOptRef = absl::optional<std::reference_wrapper<ActiveTcpListener>>;
-  using UdpListenerCallbacksOptRef =
-      absl::optional<std::reference_wrapper<Network::UdpListenerCallbacks>>;
 
-  struct ActiveListenerDetails {
+private:
+  struct ActiveTcpListenerDetail {
```
Comment (ggreenway) [inline source/server/connection_handler_impl.h:352]: I don't think these two base classes are used (at least in this PR). Delete?

File: source/server/connection_handler_impl.h [hunk 2]
```diff
@@ -364,11 +384,12 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
   UnitFloat listener_reject_fraction_{UnitFloat::min()};
 };
 
-class ActiveUdpListenerBase : public ConnectionHandlerImpl::ActiveListenerImplBase,
+using ActiveListenerImplBase = ConnectionHandlerImpl::ActiveListenerImplBase;
```
Comment (ggreenway) [inline source/server/connection_handler_impl.h:387]: I think it's easier to read without this `using`

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall

File: source/server/connection_handler_impl.h (hunk 1 above)
Comment (lambdai) [reply source/server/connection_handler_impl.h:352]: Right... deleting

File: source/server/connection_handler_impl.h (hunk 2 above)
Comment (lambdai) [reply source/server/connection_handler_impl.h:387]: ok. removed

**Review by ggreenway — APPROVED** [review]

**Comment by lambdai** [comment]
Thank you!

**Comment by lambdai** [comment]
address https://github.com/envoyproxy/envoy/pull/15275

**Pull request merged**